File: src/pso_rf/app/results.py
"""Experiment dashboard: the full 3 datasets × 5 outer folds experiment, read from ``results/<exp_id>/``.

Every number on this page comes from a saved file. The verification badge runs the same audit as
``python -m pso_rf verify`` (completeness, test isolation in the run log, summary recomputation).
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd
import streamlit as st

from pso_rf.app import charts, components
from pso_rf.app.data import DATASET_LABEL, Experiment, list_experiments
from pso_rf.experiments.verify import verify_results

REPO = Path(__file__).resolve().parents[3]


@st.cache_data(show_spinner=False)
def _verify(root: str, finished_at: str | None) -> list[str]:
    return verify_results(Path(root))


def _pick_experiment() -> Experiment | None:
    experiments = list_experiments(REPO / "results")
    if not experiments:
        st.warning("No saved experiment yet. Run `python -m pso_rf run --config configs/default.yaml`.")
        return None
    complete = [e for e in experiments if e.complete] or experiments
    with st.sidebar:
        st.subheader("Experiment")
        choice = st.selectbox(
            "Results directory",
            [e.exp_id for e in experiments],
            index=[e.exp_id for e in experiments].index(complete[0].exp_id),
        )
    return next(e for e in experiments if e.exp_id == choice)


def render() -> None:
    st.markdown(
        components.hero(
            "Experiment results",
            "3 datasets × 5 outer folds × {default RF, random search, PSO}: every sample tested exactly once, "
            "each method scored once per fold after its own search.",
        ),
        unsafe_allow_html=True,
    )
    exp = _pick_experiment()
    if exp is None:
        return
    summary, folds = exp.summary(), exp.summary_folds()
    if not exp.complete or summary.empty:
        st.info(
            f"`{exp.exp_id}` is not complete yet (status: {exp.manifest.get('status')}). "
            "Results appear here when the run finishes."
        )
        return

    _provenance(exp)
    _headline(exp, summary, folds)

    st.divider()
    tabs = st.tabs([DATASET_LABEL.get(d, d) for d in exp.datasets])
    for tab, dataset in zip(tabs, exp.datasets, strict=True):
        with tab:
            _dataset_view(exp, dataset, summary, folds)


def _provenance(exp: Experiment) -> None:
    m = exp.manifest
    problems = _verify(str(exp.root), m.get("finished_at"))
    badge = (
        components.pill("verified · all checks pass", "ok")
        if not problems
        else components.pill(f"{len(problems)} verification problem(s)", "bad")
    )
    dirty = m.get("git_dirty")
    tree = (
        components.pill("clean git tree", "ok")
        if dirty is False
        else components.pill("uncommitted changes", "bad")
    )
    st.markdown(
        f"{badge} {tree} {components.pill('config ' + m.get('config_hash', '')[:10], 'info')} "
        f"{components.pill('commit ' + (m.get('git_commit') or '?')[:8], 'info')}",
        unsafe_allow_html=True,
    )
    if problems:
        with st.expander("Verification problems"):
            for problem in problems:
                st.write("•", problem)
    st.caption(
        f"`{exp.exp_id}` · started {m.get('created_at')} · finished {m.get('finished_at')} · "
        f"Python {m.get('python_version')}, scikit-learn {m.get('packages', {}).get('scikit-learn')} · "
        f"{m.get('cpu_count')} CPUs"
    )


def _headline(exp: Experiment, summary: pd.DataFrame, folds: pd.DataFrame) -> None:
    table = summary.assign(dataset_label=summary.dataset.map(DATASET_LABEL), mean=summary.test_accuracy_mean)
    st.plotly_chart(
        charts.grouped_bars(table, "Held-out test accuracy (mean over 5 outer folds)"),
        use_container_width=True,
    )
    c1, c2 = st.columns([1.15, 1])
    deltas = folds[folds.method != "baseline"].assign(
        dataset_label=lambda f: f.dataset.map(DATASET_LABEL), delta=lambda f: f.delta_accuracy_vs_baseline
    )
    c1.plotly_chart(
        charts.delta_dots(deltas, "Paired per-fold difference to the default RF"), use_container_width=True
    )
    with c2:
        st.markdown("**Summary table** (test metrics are performance; validation is optimistically biased)")
        view = pd.DataFrame(
            {
                "dataset": summary.dataset.map(DATASET_LABEL),
                "method": summary.method.map(charts.METHOD_LABEL),
                "test acc (mean ± sd)": [
                    f"{m:.4f} ± {s:.4f}"
                    for m, s in zip(summary.test_accuracy_mean, summary.test_accuracy_std, strict=True)
                ],
                "Δ vs default": [
                    ("–" if pd.isna(d) or meth == "baseline" else f"{d:+.4f}")
                    for d, meth in zip(summary.delta_accuracy_mean, summary.method, strict=True)
                ],
                "W/T/L": [
                    ("–" if pd.isna(w) else f"{int(w)}/{int(t)}/{int(lo)}")
                    for w, t, lo in zip(summary.wins, summary.ties, summary.losses, strict=True)
                ],
                "unique fits": summary.n_unique_fits_mean.round(0).astype(int),
            }
        )
        st.dataframe(view, hide_index=True, use_container_width=True, height=360)
    st.caption(
        "No significance tests: with 5 paired folds the smallest possible two-sided Wilcoxon p-value is "
        "0.0625, so results are reported descriptively (ADR-020). Source: summary.csv, summary_folds.csv."
    )


def _dataset_view(exp: Experiment, dataset: str, summary: pd.DataFrame, folds: pd.DataFrame) -> None:
    audit = exp.audit(dataset) or {}
    rows = summary[summary.dataset == dataset].set_index("method")
    items = [
        (
            "samples",
            f"{audit.get('n_samples')}",
            f"{audit.get('n_features')} features, {audit.get('n_classes')} classes",
        )
    ]
    for method in ("baseline", "random_search", "pso"):
        if method in rows.index:
            r = rows.loc[method]
            items.append(
                (
                    charts.METHOD_LABEL[method],
                    f"{r.test_accuracy_mean:.4f}",
                    f"± {r.test_accuracy_std:.4f} test accuracy",
                )
            )
    st.markdown(components.kpis(items), unsafe_allow_html=True)

    c1, c2 = st.columns(2)
    pso_ev, rs_ev = exp.all_evaluations(dataset, "pso"), exp.all_evaluations(dataset, "random_search")
    curves, bands = {}, {}
    for method, ev in (("random_search", rs_ev), ("pso", pso_ev)):
        if ev.empty:
            continue
        grid = ev.pivot_table(index="eval_index", columns="outer_fold", values="best_so_far_fitness")
        curves[method] = grid.mean(axis=1).to_numpy()
        bands[method] = (grid.min(axis=1).to_numpy(), grid.max(axis=1).to_numpy())
    if curves:
        c1.plotly_chart(
            charts.anytime(curves, "Value of feedback: best-so-far, mean of folds (band = min–max)", bands),
            use_container_width=True,
        )
    its = exp.all_iterations(dataset)
    if not its.empty:
        mean = (
            its.groupby("iteration")
            .agg(
                gbest_fitness=("gbest_fitness", "mean"),
                mean_fitness=("mean_fitness", "mean"),
                min_fitness=("gbest_fitness", "min"),
                max_fitness=("gbest_fitness", "max"),
                gbest_n_estimators=("gbest_n_estimators", "median"),
                gbest_max_depth=("gbest_max_depth", "median"),
                gbest_min_samples_split=("gbest_min_samples_split", "median"),
            )
            .reset_index()
        )
        c2.plotly_chart(
            charts.convergence(
                mean.to_dict("records"), "PSO convergence, mean of folds (band = fold range of gbest)"
            ),
            use_container_width=True,
        )

    st.markdown("**Chosen configurations per outer fold** (ties with the best are common on flat landscapes)")
    sub = folds[folds.dataset == dataset]
    chosen = []
    for row in sub.itertuples():
        final = exp.final(dataset, int(row.outer_fold), row.method) or {}
        chosen.append(
            {
                "fold": row.outer_fold,
                "method": charts.METHOD_LABEL[row.method],
                "config (n, depth, split)": components.config_text(final.get("best_hyperparameters")),
                "validation (biased)": row.validation_fitness,
                "test accuracy": row.test_accuracy,
                "Δ vs default": row.delta_accuracy_vs_baseline,
                "ties with best": final.get("n_ties_with_best"),
                "on a bound": ", ".join(final.get("boundary_hits") or []) or "–",
            }
        )
    st.dataframe(
        pd.DataFrame(chosen).style.format(
            {"validation (biased)": "{:.4f}", "test accuracy": "{:.4f}", "Δ vs default": "{:+.4f}"},
            na_rep="–",
        ),
        hide_index=True,
        use_container_width=True,
    )

    c3, c4 = st.columns(2)
    labels = audit.get("class_names") or []
    for column, method in ((c3, "baseline"), (c4, "pso")):
        pred = exp.predictions(dataset, method)
        if pred.empty or not labels:
            continue
        n = len(labels)
        matrix = np.zeros((n, n), dtype=int)
        np.add.at(matrix, (pred.y_true.to_numpy(), pred.y_pred.to_numpy()), 1)
        names = labels if n <= 3 else [str(i) for i in range(n)]
        column.plotly_chart(
            charts.confusion(matrix, names, f"Pooled confusion · {charts.METHOD_LABEL[method]}"),
            use_container_width=True,
        )

    dep = exp.deployment(dataset)
    if dep:
        est = dep.get("performance_estimate") or {}
        st.success(
            f"**Recommended configuration (deployment run on all data):** "
            f"{components.config_text(dep['recommended_hyperparameters'])} · expected test accuracy "
            f"≈ {est.get('test_accuracy_mean', float('nan')):.4f} ± {est.get('test_accuracy_std', float('nan')):.4f} "
            f"(outer-CV estimate of the same procedure; the deployment run itself has no test data)."
        )

File: src/pso_rf/web/static/js/views/results.js
// Results explorer: the saved 3 datasets × 5 outer folds × 3 methods experiment (files only).
import { api } from "../api.js";
import { LineChart, barChart, heatmap } from "../charts.js";
import { COLORS, LABEL, fmt, h, kpi } from "../ui.js";

const METHODS = ["baseline", "random_search", "pso"];

export function mount(root, meta) {
  const charts = [], panelCharts = [];
  let disposed = false;
  const body = h("div", { class: "grid" }, h("div", { class: "caption" }, "Loading saved results…"));
  const expSel = h("select", {}, ...meta.experiments.map((e) => h("option", { value: e.id }, `${e.id} · ${e.status}`)));
  if (meta.default_experiment) expSel.value = meta.default_experiment;
  expSel.addEventListener("change", () => load(expSel.value));
  root.append(h("div", { class: "view-head" }, h("div", {}, h("h1", {}, "Experiment results"),
    h("p", {}, "3 datasets × 5 outer folds × {default RF, random search, PSO}. Every sample is tested exactly once, and each method is scored once per fold after its own search. Everything on this page is read from saved result files.")),
    h("div", { class: "ctl" }, h("label", {}, "Results directory"), expSel)), body);

  async function load(exp) {
    [...charts.splice(0), ...panelCharts.splice(0)].forEach((c) => c.destroy && c.destroy());
    body.replaceChildren(h("div", { class: "caption" }, "Loading saved results…"));
    const R = await api.results(exp);
    if (disposed) return;
    body.replaceChildren(provenance(R), findings(R), overview(R), datasets(R));
  }

  function provenance(R) {
    const m = R.manifest, ok = !R.problems.length;
    return h("div", { class: "card" }, h("div", { class: "btn-row", style: { alignItems: "center" } },
      h("span", { class: `pill ${ok ? "ok" : "bad"}` }, ok ? "✓ verified: all checks pass" : `✕ ${R.problems.length} verification problem(s)`),
      h("span", { class: `pill ${m.git_dirty === false ? "ok" : "bad"}` }, m.git_dirty === false ? "✓ clean git tree" : "uncommitted changes"),
      h("span", { class: "pill info" }, `commit ${(m.git_commit || "?").slice(0, 8)}`),
      h("span", { class: "pill info" }, `config ${m.config_hash.slice(0, 10)}`),
      h("span", { class: "caption" }, `${m.created_at} → ${m.finished_at} · Python ${m.python_version} · scikit-learn ${m.packages["scikit-learn"]} · ${m.cpu_count} CPUs`)),
      ok ? null : h("ul", {}, ...R.problems.map((p) => h("li", {}, p))),
      h("div", { class: "caption", style: { marginTop: "8px" } }, "The badge re-runs the same audit as python -m pso_rf verify: completeness, test isolation in the run log, and summary recomputation."));
  }

  function row(R, ds, method) { return R.summary.find((r) => r.dataset === ds && r.method === method); }

  function findings(R) {
    const items = Object.keys(R.datasets).map((ds) => {
      const pso = row(R, ds, "pso"), base = row(R, ds, "baseline"), rs = row(R, ds, "random_search");
      const nTest = R.datasets[ds].audit.n_samples / 5, one = 1 / nTest;
      const d = pso.delta_accuracy_mean, verdict = Math.abs(d) < one ? "within one test sample of" : d > 0 ? "above" : "below";
      return h("div", { class: "kpi" }, h("div", { class: "k" }, R.datasets[ds].label),
        h("div", { class: "v", style: { fontSize: "18px" } }, `PSO ${fmt.acc(pso.test_accuracy_mean)}`),
        h("div", { class: "s" }, `default ${fmt.acc(base.test_accuracy_mean)} · random ${fmt.acc(rs.test_accuracy_mean)}`),
        h("div", { class: "s", style: { marginTop: "4px" } }, `PSO is ${verdict} the default (Δ ${fmt.delta(d)}; W/T/L ${pso.wins}/${pso.ties}/${pso.losses}); vs random search ${pso.wins_vs_rs}/${pso.ties_vs_rs}/${pso.losses_vs_rs}.`),
        h("div", { class: "s" }, `Unique RF fits: PSO ${pso.n_unique_fits_mean.toFixed(0)} vs random ${rs.n_unique_fits_mean.toFixed(0)} per fold.`));
    });
    return h("div", { class: "grid" }, h("div", { class: "kpis", style: { gridTemplateColumns: "repeat(auto-fit, minmax(300px, 1fr))" } }, ...items),
      h("div", { class: "insight" }, h("b", {}, "How to read this. "), "Only held-out test accuracy is performance. Validation accuracy is the best of many noisy estimates, so it is optimistically biased. With 5 paired folds no significance test is possible (the smallest two-sided Wilcoxon p-value is 0.0625), so the results are descriptive. A difference smaller than one test sample is noise."));
  }

  function overview(R) {
    const box = h("div");
    const groups = Object.keys(R.datasets).map((ds) => ({
      label: R.datasets[ds].label,
      bars: METHODS.map((m) => ({ key: m, label: LABEL[m], color: COLORS[m], value: row(R, ds, m).test_accuracy_mean,
        dots: R.folds.filter((f) => f.dataset === ds && f.method === m).map((f) => f.test_accuracy) })),
    }));
    const card = h("div", { class: "card" }, h("h3", {}, "Held-out test accuracy", h("span", { class: "sub" }, "bar = mean of 5 outer folds, dots = folds; hover for values")), box,
      h("div", { class: "legend" }, ...METHODS.map((m) => h("span", {}, h("i", { style: { background: COLORS[m], height: "10px", width: "10px" } }), LABEL[m]))));
    requestAnimationFrame(() => charts.push(barChart(box, groups, { height: 300, yLabel: "test accuracy" })));
    return card;
  }

  function datasets(R) {
    const keys = Object.keys(R.datasets), tabs = h("div", { class: "tabs", role: "tablist" }), panel = h("div");
    const show = (ds) => {
      [...tabs.children].forEach((b) => b.classList.toggle("on", b.dataset.ds === ds));
      panelCharts.splice(0).forEach((c) => c.destroy && c.destroy());
      panel.replaceChildren(datasetPanel(R, ds));
    };
    keys.forEach((ds) => { const b = h("button", { type: "button", role: "tab", "data-ds": ds }, R.datasets[ds].label); b.addEventListener("click", () => show(ds)); tabs.append(b); });
    requestAnimationFrame(() => show(keys[keys.length - 1]));
    return h("div", { class: "card" }, tabs, panel);
  }

  function datasetPanel(R, ds) {
    const D = R.datasets[ds];
    const anyBox = h("div"), convBox = h("div");
    const wrap = h("div", { class: "grid" });
    const dep = D.deployment;
    wrap.append(h("div", { class: "kpis" },
      kpi("samples", D.audit.n_samples, `${D.audit.n_features} features · ${D.audit.n_classes} classes`),
      kpi("class balance", D.audit.class_ratio.toFixed(2), "max/min class ratio"),
      kpi("fitness metric", D.audit.fitness_metric, "inner 5-fold CV"),
      dep ? kpi("recommended config", fmt.cfg(dep.recommended_hyperparameters), "deployment run, all data") : null,
      dep && dep.performance_estimate ? kpi("expected test accuracy", fmt.acc(dep.performance_estimate.test_accuracy_mean),
        `± ${fmt.acc(dep.performance_estimate.test_accuracy_std)} (outer-CV estimate)`) : null));
    wrap.append(h("div", { class: "grid cols-2" },
      h("div", {}, h("h3", {}, "Value of feedback", h("span", { class: "sub" }, "best validation accuracy so far · mean of folds, band = min–max")), anyBox),
      h("div", {}, h("h3", {}, "PSO convergence", h("span", { class: "sub" }, "gbest per iteration, one line per outer fold")), convBox)));
    requestAnimationFrame(() => {
      const any = new LineChart(anyBox, { xLabel: "configurations evaluated", yLabel: "best so far", height: 260 });
      for (const m of ["random_search", "pso"]) {
        const a = D.anytime[m]; if (!a) continue;
        any.set(`${m}-band`, { color: COLORS[m], band: a.mean.map((_, i) => [i + 1, a.min[i], a.max[i]]) });
        any.set(m, { color: COLORS[m], points: a.mean.map((v, i) => [i + 1, v]), step: true, width: 2.5, label: LABEL[m] });
      }
      const conv = new LineChart(convBox, { xLabel: "PSO iteration", yLabel: "gbest validation accuracy", integerX: true, height: 260 });
      D.convergence.forEach((c) => conv.set(`f${c.fold}`, { color: COLORS.pso, points: c.gbest.map((v, i) => [i, v]), step: true, width: 1.4, opacity: 0.45, label: `fold ${c.fold}` }));
      const n = D.convergence[0]?.gbest.length || 0;
      conv.set("mean", { color: "#ffffff", points: Array.from({ length: n }, (_, i) => [i, D.convergence.reduce((s, c) => s + c.gbest[i], 0) / D.convergence.length]), step: true, width: 2.5, label: "mean of folds" });
      panelCharts.push(any, conv);
    });
    const tbl = h("table", { class: "t" }, h("thead", {}, h("tr", {}, h("th", {}, "fold"), h("th", {}, "method"), h("th", {}, "configuration"),
      h("th", { class: "num" }, "validation (biased)"), h("th", { class: "num" }, "test"), h("th", { class: "num" }, "ties with best"), h("th", {}, "on a bound"))),
      h("tbody", {}, ...D.chosen.sort((a, b) => a.fold - b.fold || METHODS.indexOf(a.method) - METHODS.indexOf(b.method)).map((c) => h("tr", {},
        h("td", {}, c.fold), h("td", {}, h("span", { class: "sw", style: { background: COLORS[c.method] } }), LABEL[c.method]),
        h("td", { class: "mono" }, fmt.cfg(c.cfg)), h("td", { class: "num" }, fmt.acc(c.val)), h("td", { class: "num" }, fmt.acc(c.test)),
        h("td", { class: "num" }, c.ties ?? "–"), h("td", {}, (c.bounds || []).join(", ") || "–")))));
    wrap.append(h("div", {}, h("h3", {}, "Chosen configuration per fold", h("span", { class: "sub" }, "many configurations tie at the top: flat landscapes")), h("div", { style: { overflowX: "auto" } }, tbl)));
    const cms = h("div", { class: "grid cols-2" });
    for (const m of ["baseline", "pso"]) {
      const folds = D.chosen.filter((c) => c.method === m);
      if (!folds.length) continue;
      const sum = folds[0].cm.map((r, i) => r.map((_, j) => folds.reduce((s, f) => s + f.cm[i][j], 0)));
      const box = h("div"); heatmap(box, sum, D.audit.class_names, { title: `${LABEL[m]} · pooled over 5 folds (rows = true)` });
      cms.append(box);
    }
    wrap.append(h("div", {}, h("h3", {}, "Pooled confusion matrices", h("span", { class: "sub" }, "every sample tested once · blue = correct, red = errors")), cms));
    return wrap;
  }

  load(expSel.value).catch((err) => body.replaceChildren(h("div", { class: "card" }, String(err.message || err))));
  return () => { disposed = true; [...charts, ...panelCharts].forEach((c) => c.destroy && c.destroy()); };
}
